File: src/main/java/org/jcodec/movtool/QTEdit.java
package org.jcodec.movtool;
import java.lang.IllegalStateException;
import java.lang.System;


import org.jcodec.containers.mp4.boxes.MovieBox;
import org.jcodec.movtool.Flattern.ProgressListener;

import java.io.File;
import java.nio.channels.FileChannel;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.LinkedList;
import java.util.List;

/**
 * This class is part of JCodec ( www.jcodec.org ) This software is distributed
 * under FreeBSD License
 * 
 * @author The JCodec project
 * 
 */
public class QTEdit {

    protected final EditFactory[] factories;
    private final List<ProgressListener> listeners;

    public static interface EditFactory {
        String getName();

        MP4Edit parseArgs(List<String> args);

        String getHelp();
    }

    public static abstract class BaseCommand implements MP4Edit {
        public void applyRefs(MovieBox movie, FileChannel[][] refs) {
            apply(movie);
        }

        public abstract void apply(MovieBox movie);
    }

    public QTEdit(EditFactory... arguments) {
        this.listeners = new ArrayList<ProgressListener>();
        this.factories = arguments;
    }

    public void addProgressListener(ProgressListener listener) {
        listeners.add(listener);
    }

    public void execute(String[] args) throws Exception {
        LinkedList<String> aa = new LinkedList<String>(Arrays.asList(args));

        final List<MP4Edit> commands = new LinkedList<MP4Edit>();
        while (aa.size() > 0) {
            int i;
            for (i = 0; i < factories.length; i++) {
                if (aa.get(0).equals(factories[i].getName())) {
                    aa.remove(0);
                    try {
                        commands.add(factories[i].parseArgs(aa));
                    } catch (Exception e) {
                        System.err.println("ERROR: " + e.getMessage());
                        return;
                    }
                    break;
                }
            }
            if (i == factories.length)
                break;
        }
        if (aa.size() == 0) {
            System.err.println("ERROR: A movie file should be specified");
            help();
        }
        if (commands.size() == 0) {
            System.err.println("ERROR: At least one command should be specified");
            help();
        }
        File input = new File(aa.remove(0));

        if (!input.exists()) {
            System.err.println("ERROR: Input file '" + input.getAbsolutePath() + "' doesn't exist");
            help();
        }

        new ReplaceMP4Editor().replace(input, new CompoundMP4Edit(commands));
    }

    protected void help() {
        System.out.println("Quicktime movie editor");
        System.out.println("Syntax: qtedit <command1> <options> ... <commandN> <options> <movie>");
        System.out.println("Where options:");
        for (EditFactory commandFactory : factories) {
            System.out.println("\t" + commandFactory.getHelp());
        }

        System.exit(-1);
    }
}

File: src/main/java/org/jcodec/movtool/QTEdit.kt
package org.jcodec.movtool

import org.jcodec.containers.mp4.boxes.MovieBox
import java.io.File
import java.nio.channels.FileChannel
import java.util.*

/**
 * This class is part of JCodec ( www.jcodec.org ) This software is distributed
 * under FreeBSD License
 *
 * @author The JCodec project
 */
class QTEdit(editFactories: Array<EditFactory>) {
    protected val factories: Array<EditFactory>
    private val listeners: MutableList<Flatten.ProgressListener>

    interface EditFactory {
        val name: String
        fun parseArgs(args: List<String>?): MP4Edit?
        val help: String
    }

    abstract class BaseCommand : MP4Edit {
        fun applyRefs(movie: MovieBox, refs: Array<Array<FileChannel?>?>?) {
            apply(movie)
        }

        abstract override fun apply(movie: MovieBox)
    }

    fun addProgressListener(listener: Flatten.ProgressListener) {
        listeners.add(listener)
    }

    @Throws(Exception::class)
    fun execute(args: Array<String>) {
        val aa = LinkedList(Arrays.asList(*args))
        val commands: MutableList<MP4Edit?> = LinkedList()
        while (aa.size > 0) {
            var i: Int
            i = 0
            while (i < factories.size) {
                if (aa[0] == factories[i].name) {
                    aa.removeAt(0)
                    try {
                        commands.add(factories[i].parseArgs(aa))
                    } catch (e: Exception) {
                        System.err.println("ERROR: " + e.message)
                        return
                    }
                    break
                }
                i++
            }
            if (i == factories.size) break
        }
        if (aa.size == 0) {
            System.err.println("ERROR: A movie file should be specified")
            help()
        }
        if (commands.size == 0) {
            System.err.println("ERROR: At least one command should be specified")
            help()
        }
        val input = File(aa.removeAt(0))
        if (!input.exists()) {
            System.err.println("ERROR: Input file '" + input.absolutePath + "' doesn't exist")
            help()
        }
        ReplaceMP4Editor().replace(input, CompoundMP4Edit(commands))
    }

    protected fun help() {
        println("Quicktime movie editor")
        println("Syntax: qtedit <command1> <options> ... <commandN> <options> <movie>")
        println("Where options:")
        for (commandFactory in factories) {
            println("\t" + commandFactory.help)
        }
        System.exit(-1)
    }

    init {
        listeners = ArrayList()
        factories = editFactories
    }
}